File: FunctionalPatterns/Structural/Adapter.fs
namespace FunctionalPatterns
namespace FunctionalPatterns.Structural

module Adapter =

    type Greetings =
        { greetFormal: unit -> string
          greetInformal: unit -> string }

    type Hello =
        { sayHello: string -> unit
          whatsUp: string -> unit }

    let newHello name =
        { sayHello = fun prefix -> printfn "Hello %s %s" prefix name
          whatsUp = fun prefix -> printfn "What's up %s %s" prefix name}

    // Type to be adapted: Hello
    // Type that is needed: Greetings
    // Adapter pattern is basically a function:
    //  Hello -> <Additional Hello config (here prefix)> -> Greetings
    let newGreetingsAdapter legacyHello prefix =
        { greetFormal = fun () -> legacyHello.sayHello prefix; prefix
          greetInformal = fun () -> legacyHello.whatsUp prefix; prefix }

File: ObjectOrientedPatterns/Structural/Adapter.cs
namespace ObjectOrientedPatterns.Structural.Adapter
{
    // see https://en.wikipedia.org/wiki/Adapter_pattern

    public interface IGreetings
    {
        public string GreetFormal();
        public string GreetInformal();
    }

    public class SayHelloAdapter : IGreetings
    {
        public SayHelloAdapter(SayHello sayHello, string prefix)
        {
            mySayHello = sayHello;
            myPrefix = prefix;
        }
        public string GreetFormal()
        {
            mySayHello.Hello(myPrefix);
            return myPrefix;
        }
        public string GreetInformal()
        {
            mySayHello.WhatsUp(myPrefix);
            return myPrefix;
        }
        private readonly SayHello mySayHello;
        private readonly string myPrefix;
    }

    public class SayHello
    {
        public SayHello(string name)
        {
            myName = name;
        }
        public void Hello(string prefix)
        {
            Console.WriteLine("Hello " + prefix + " " + myName);
        }
        public void WhatsUp(string prefix)
        {
            Console.WriteLine("What's up " + prefix + " " + myName);
        }
        private readonly string myName;
    }
}
